src/test/java/RegistrationTests.java

```diff
@@ -0,0 +1,36 @@
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class RegistrationTests extends BaseTest{
+
+    public void successfulRegistrationMessage() {
+        String messageExpected = "We've sent a confirmation link to the email. Please continue by clicking on it";
+        WebElement messageActual = driver.findElement(By.xpath("//div[@class='login-wrapper']//div[@class='messages']"));
+        String message = messageActual.getText();
+        System.out.println(message);
+        Assert.assertEquals(message, messageExpected);
+    }
+    @Test
+    public void registrationNavigation() throws InterruptedException {
+        //Step1: Click on the Registration link and assert that the correct page is opened
+        WebElement registrationLink = driver.findElement(By.cssSelector("[href=\"registration\"]"));
+        registrationLink.click();
+        Assert.assertEquals(driver.getCurrentUrl(), "https://qa.koel.app/registration");
+
+        //Step2: Access the email field and enter a valid email
+        //It only accepts testpro.io emails
+        WebElement emailField = driver.findElement(By.cssSelector("[type=\"email\"]"));
+        emailField.clear();
+        emailField.sendKeys("abcd@testpro.io");
+
+        //Step3: Click on the submit button
+        WebElement submitButton = driver.findElement(By.cssSelector("[id=\"button\"]"));
+        submitButton.click();
+        Thread.sleep(2000);
+
+        //Step4: Assert that the message text is displayed using contains()
+        successfulRegistrationMessage();
+    }
```

Comment:
1. **BasePage and BaseTest** 

**BaseTest** — Implements the necessary methods for work with webdriver.

The BaseTest class contains common setup and teardown methods, as well as any shared functionality or configurations required by multiple test classes.

**BaseTest**

The problem we are trying to solve here is the reuse of common behaviors across test classes avoiding coding duplication and centralizing these actions at one point.

**@BeforeSuite**	- The annotated method will be run before all tests in this suite have run.

**@BeforeMethod** -	The annotated method will be run before all the test methods in the current class have been run.

**@AfterMethod**	- The annotated method will be run after each test method.

_Example without a Base Test Class_

![image](https://github.com/Test-Pro-Bootcamp/A53-QA-Automation-Framework/assets/104761482/c3abad8a-8bc1-4ad0-8a66-23bd19e7af48)


**BasePage**  — class serves as a foundation or parent class for individual page classes. It contains common methods and functionalities that are shared across multiple pages in an application.

_Example without a Base Page Class_

![image](https://github.com/Test-Pro-Bootcamp/A53-QA-Automation-Framework/assets/104761482/4a96b82c-8bc8-4ce6-927d-87b7e248a849)


2. **Test in Page Object Model (POM)**

![Clean_Test](https://user-images.githubusercontent.com/104761482/236863813-f30d7bfc-e89d-4419-bdea-4b1517155e39.png)

3.  **Page Object Model and Page Factory**

![POM_vs_Factory_1](https://user-images.githubusercontent.com/104761482/236864030-34b89652-460d-4440-aa91-894332e90654.png)

![POM_vs_Factory_2](https://user-images.githubusercontent.com/104761482/236864098-d3ffe636-7ba9-484e-a72f-8b03ff024659.png)

4.  Please, use **Encapsulation**

Encapsulation is one of the four fundamental OOP concepts. The other three are inheritance, polymorphism, and abstraction.

You should use encapsulation of non-static and non-final fields.

If a variable is used only by other methods of that class, it should be declared as private.

Make all the locators and WebElements private, because private keyword in java allows most restrictive access to variables and methods and offer strongest form of Encapsulation. 

For example

    ```
    private By shuffleBtnLocator = By.cssSelector(".btn-shuffle-all");
    private By firstSongLocator = By.cssSelector(".play");

```
   @FindBy(css = "img.avatar")
    private WebElement avatar
```


Make the variables protected, because protected access gives the subclass a chance to use the helper method or variable, while preventing a nonrelated class from trying to use it.

```
    protected Actions actions;
    protected WebDriverWait wait;
    protected WebDriver driver;
```